README.md
openmediavault-nginx
====================

Nginx plugin for OpenMediaVault


debian/changelog
openmediavault-nginx (4.0.4) stable; urgency=low

  * Fix for change in OMV build_path behavior

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 29 Jan 2020 07:21:32 -0600

openmediavault-nginx (4.0.3) stable; urgency=low

  * Add http/2 support

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 24 Sep 2018 09:30:13 -0500

openmediavault-nginx (4.0.2) stable; urgency=low

  * Revert 4.0.1 changes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 21 Jan 2018 09:29:18 -0600

openmediavault-nginx (4.0.1) stable; urgency=low

  * Remove ipv4 listen directive on ipv6 enabled hosts

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 19 Jan 2018 08:32:02 -0600

openmediavault-nginx (4.0) stable; urgency=low

  * Update for OMV 4.x

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 02 Sep 2017 12:18:16 -0500

openmediavault-nginx (3.0.11) stable; urgency=low

  * Improve foreign content handling

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 18 Mar 2017 10:40:47 -0500

openmediavault-nginx (3.0.10) stable; urgency=low

  * Fix upgrades

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 24 Jan 2017 16:54:41 -0600

openmediavault-nginx (3.0.9) stable; urgency=medium

  * Improve the behaviour of the extra options field when editing a server.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 15 Nov 2016 23:02:17 +0100

openmediavault-nginx (3.0.8) stable; urgency=medium

  * Load the plugin configuration after the one from OMV so that the web
    interface is inexplicitly the default server.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 04 Nov 2016 18:00:54 +0100

openmediavault-nginx (3.0.7) stable; urgency=medium

  * Fix typo in mkconf script.
  * Fix incorrectly written arithmetric operation so that configuration
    generation doesn't fail.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 28 Sep 2016 18:35:45 +0200

openmediavault-nginx (3.0.6) stable; urgency=medium

  * Make sure both IPv4 and IPv6 work.
  * OMV 3.0.40 compatibility.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 27 Sep 2016 21:06:51 +0200

openmediavault-nginx (3.0.5) stable; urgency=low

  * Fix default server setting

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 22 Jul 2016 18:31:13 -0500

openmediavault-nginx (3.0.4) stable; urgency=medium

  * OMV 3.0.21 compatibility.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 04 Jun 2016 19:08:18 +0200

openmediavault-nginx (3.0.3) stable; urgency=medium

  * OMV 3.0.15 support.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 07 Apr 2016 15:42:29 +0200

openmediavault-nginx (3.0.2) stable; urgency=medium

  * Use snippets/fastcgi-php.conf as the default PHP config in Nginx.
  * Minor web interface tweaks.
  * The user and group openmediavault is now not longer available as an
    option for a FPM pool.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 27 Jan 2016 23:40:14 +0100

openmediavault-nginx (3.0.1) stable; urgency=medium

  * Improve configuration script.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 01 Jan 2016 23:25:07 +0100

openmediavault-nginx (3.0) stable; urgency=low

  * OMV 3.0 release

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 30 Dec 2015 14:24:31 +0100

openmediavault-nginx (2.3) stable; urgency=low

  * Fixed issue where Nginx got misconfigured and wouldn't start when no index
    page was selected but the use index was still activated.
  * Utilize helper in OMV 2.1.9.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Thu, 15 Oct 2015 23:43:31 +0200

openmediavault-nginx (2.2) stable; urgency=low

  * Fix cert location

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 03 Jul 2015 16:45:50 -0500

openmediavault-nginx (2.1) stable; urgency=low

  * Update ssl certificate combo for OMV 2.x
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sun, 28 Jun 2015 19:58:44 -0500

openmediavault-nginx (1.4) stable; urgency=medium

  * Add IPv6 support.
  * Allow using the same port as the web interface when using name-based
    servers.
  * Fix issue which caused omv-mkconf to fail when enabled servers was
    preceded by a disabled server in config.xml.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 03 Apr 2015 21:12:18 +0200

openmediavault-nginx (1.1) stable; urgency=low

  * Improve port validation.

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 28 Jan 2015 19:10:58 +0100

openmediavault-nginx (1.0.14) stable; urgency=low

  * Change icon title
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 26 Jan 2015 13:28:56 -0600

openmediavault-nginx (1.0.13) stable; urgency=low

  * Add custom debian/control field XB-Plugin-Section
  * Update locales

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Fri, 12 Dec 2014 06:11:50 -0600

openmediavault-nginx (1.0.12) stable; urgency=low

  * Resolve #4

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Tue, 04 Nov 2014 12:11:59 +0100

openmediavault-nginx (1.0.11) stable; urgency=low

  * Reallow renaming of pool

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Mon, 03 Nov 2014 14:44:06 +0100

openmediavault-nginx (1.0.10) unstable; urgency=low

  * Disallow removal of FPM pool in use
  * Disallow renaming pool
  * Various fixes

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Wed, 29 Oct 2014 15:12:58 +0100

openmediavault-nginx (1.0.9) unstable; urgency=low

  * Fix error in mkconf causing multiple default servers on same port

 -- OpenMediaVault Plugin Developers <plugins@omv-extras.org>  Sat, 26 Jul 2014 15:58:47 +0200

openmediavault-nginx (1.0.8) unstable; urgency=low


var/www/openmediavault/js/omv/module/admin/service/nginx/Nginx.js
/**
 * Copyright (C) 2014-2017 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")

OMV.WorkspaceManager.registerNode({
    id: 'nginx',
    path: '/service',
    text: _('Nginx (websites)'),
    icon16: 'images/nginx.png',
    iconSvg: 'images/nginx.svg'
});


var/www/openmediavault/js/omv/module/admin/service/nginx/Pools.js
/**
 * Copyright (C) 2014-2017 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/module/admin/service/nginx/window/Pool.js")

Ext.define('OMV.module.admin.service.nginx.Pools', {
    extend: 'OMV.workspace.grid.Panel',
    requires: [
        'OMV.data.Store',
        'OMV.data.Model',
        'OMV.data.proxy.Rpc',
        'OMV.module.admin.service.nginx.window.Pool'
    ],

    hidePagingToolbar: false,
    reloadOnActivate: true,

    columns: [{
        xtype: "textcolumn",
        header: _('UUID'),
        hidden: true,
        dataIndex: 'uuid'
    }, {
        xtype: "textcolumn",
        header: _('Name'),
        flex: 1,
        sortable: true,
        dataIndex: 'name'
    }, {
        xtype: "textcolumn",
        header: _('Description'),
        flex: 1,
        sortable: true,
        dataIndex: 'description'
    }],

    store: Ext.create('OMV.data.Store', {
        autoLoad: true,
        model: OMV.data.Model.createImplicit({
            idProperty: 'uuid',
            fields: [{
                name: 'uuid'
            }, {
                name: 'name'
            }, {
                name: 'description'
            }]
        }),
        proxy: {
            type: 'rpc',
            rpcData: {
                'service': 'PhpFpm',
                'method': 'getList'
            }
        },
        remoteSort: true,
        sorters: [{
            direction: 'ASC',
            property: 'name'
        }]
    }),

    onAddButton: function() {
        Ext.create('OMV.module.admin.service.nginx.window.Pool', {
            title: _('Add pool'),
            uuid: OMV.UUID_UNDEFINED,
            listeners: {
                scope: this,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton: function() {
        var record = this.getSelected();

        Ext.create('OMV.module.admin.service.nginx.window.Pool', {
            rpcGetMethod: 'get',
            title: _('Edit pool'),
            uuid: record.get('uuid'),
            listeners: {
                scope: this,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion: function(record) {
        OMV.Rpc.request({
            callback: this.onDeletion,
            rpcData: {
                service: 'PhpFpm',
                method: 'delete',
                params: {
                    uuid: record.get('uuid')
                }
            },
            scope: this
        });
    }

});

OMV.WorkspaceManager.registerPanel({
    id: 'pools',
    path: '/service/nginx',
    text: _('Pools'),
    position: 30,
    className: 'OMV.module.admin.service.nginx.Pools'
});


var/www/openmediavault/js/omv/module/admin/service/nginx/Servers.js
/**
 * Copyright (C) 2014-2017 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/grid/Panel.js")
// require("js/omv/data/Model.js")
// require("js/omv/data/proxy/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/module/admin/service/nginx/window/Server.js")
// require("js/omv/module/admin/service/nginx/window/Log.js")

Ext.define('OMV.module.admin.service.nginx.Servers', {
    extend: 'OMV.workspace.grid.Panel',
    requires: [
        'OMV.data.Model',
        'OMV.data.proxy.Rpc',
        'OMV.data.Store',
        'OMV.module.admin.service.nginx.window.Log',
        'OMV.module.admin.service.nginx.window.Server'
    ],

    hidePagingToolbar: false,
    accessLogButtonText: _('Access log'),
    errorLogButtonText: _('Error log'),
    reloadOnActivate: true,

    columns: [{
        xtype: "textcolumn",
        header: _('UUID'),
        hidden: true,
        dataIndex: 'uuid'
    }, {
        xtype: 'booleaniconcolumn',
        header: _('Enabled'),
        sortable: true,
        dataIndex: 'enable',
        align: 'center',
        width: 80,
        resizable: false,
        trueIcon: 'switch_on.png',
        falseIcon: 'switch_off.png'
    }, {
        xtype: 'booleaniconcolumn',
        header: 'SSL',
        sortable: true,
        dataIndex: 'ssl_enable',
        align: 'center',
        width: 60,
        resizable: false,
        trueIcon: 'switch_on.png',
        falseIcon: 'switch_off.png'
    }, {
        xtype: 'booleaniconcolumn',
        header: 'PHP',
        sortable: true,
        dataIndex: 'php_enable',
        align: 'center',
        width: 60,
        resizable: false,
        trueIcon: 'switch_on.png',
        falseIcon: 'switch_off.png'
    }, {
        xtype: 'booleaniconcolumn',
        header: _('Default'),
        sortable: true,
        dataIndex: 'port_default_server',
        align: 'center',
        width: 60,
        resizable: false,
        trueIcon: 'switch_on.png',
        falseIcon: 'switch_off.png'
    }, {
        xtype: 'booleaniconcolumn',
        header: _('Default (SSL)'),
        sortable: true,
        dataIndex: 'ssl_port_default_server',
        align: 'center',
        width: 80,
        resizable: false,
        trueIcon: 'switch_on.png',
        falseIcon: 'switch_off.png'
    }, {
        xtype: "textcolumn",
        header: _('Root'),
        flex: 1,
        sortable: true,
        dataIndex: 'root_full_path'
    }, {
        xtype: "textcolumn",
        header: _('URL'),
        flex: 1,
        sortable: true,
        dataIndex: 'urls',
        renderer: function(urls) {
            var value = '';

            for (var i = 0; i < urls.length; i++)
                urls[i] = Ext.String.format('<a href="{0}" target="_blank">{0}</a>', urls[i]
                    .replace('!domain!', window.location.hostname));

            value = urls.join(', ');

            return value;
        }
    }],

    store: Ext.create('OMV.data.Store', {
        autoLoad: true,
        model: OMV.data.Model.createImplicit({
            idProperty: 'uuid',
            fields: [{
                name: 'uuid'
            }, {
                name: 'enable'
            }, {
                name: 'log_enable'
            }, {
                name: 'ssl_enable'
            }, {
                name: 'php_enable'
            }, {
                name: 'port_default_server'
            }, {
                name: 'ssl_port_default_server'
            }, {
                name: 'root_full_path'
            }, {
                name: 'urls'
            }]
        }),
        proxy: {
            type: 'rpc',
            rpcData: {
                'service': 'Nginx',
                'method': 'getList'
            }
        },
        remoteSort: true
    }),

    getTopToolbarItems: function() {
        var items = this.callParent(arguments);

        Ext.Array.push(items, [{
            xtype: 'tbseparator'
        }, {
            id: this.getId() + '-access-log',
            xtype: 'button',
            text: this.accessLogButtonText,
            handler: Ext.Function.bind(this.onViewLogButton, this, ['access']),
            scope: this,
            disabled: true,
            selectionConfig: {
                minSelections: 1,
                maxSelections: 1,
                enabledFn: Ext.Function.bind(this.logButtonEnabled, this)
            }
        }, {
            id: this.getId() + '-error-log',
            xtype: 'button',
            text: this.errorLogButtonText,
            handler: Ext.Function.bind(this.onViewLogButton, this, ['error']),
            scope: this,
            disabled: true,
            selectionConfig: {
                minSelections: 1,
                maxSelections: 1,
                enabledFn: Ext.Function.bind(this.logButtonEnabled, this)
            }
        }]);

        return items;
    },

    logButtonEnabled: function(button, records) {
        var record = records[0];

        return record.get('log_enable');
    },

    onAddButton: function() {
        Ext.create('OMV.module.admin.service.nginx.window.Server', {
            title: _('Add server'),
            uuid: OMV.UUID_UNDEFINED,
            listeners: {
                scope: this,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    onEditButton: function() {
        var record = this.getSelected();

        Ext.create('OMV.module.admin.service.nginx.window.Server', {
            rpcGetMethod: 'get',
            title: _('Edit server'),
            uuid: record.get('uuid'),
            listeners: {
                scope: this,
                submit: function() {
                    this.doReload();
                }
            }
        }).show();
    },

    doDeletion: function(record) {
        OMV.Rpc.request({
            scope: this,
            callback: this.onDeletion,
            rpcData: {
                service: 'Nginx',
                method: 'delete',
                params: {
                    uuid: record.get('uuid')
                }
            }
        });
    },

    onViewLogButton: function(logType) {
        var record = this.getSelected();

        Ext.create('OMV.module.admin.service.nginx.window.Log', {
            uuid: record.get('uuid'),
            logType: logType
        }).show();
    }
});

OMV.WorkspaceManager.registerPanel({
    id: 'servers',
    path: '/service/nginx',
    text: _('Servers'),
    position: 20,
    className: 'OMV.module.admin.service.nginx.Servers'
});


var/www/openmediavault/js/omv/module/admin/diagnostic/log/plugin/PhpFpm.js
/**
 * Copyright (C) 2014 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/PluginManager.js")
// require("js/omv/module/admin/diagnostic/log/plugin/Plugin.js")

Ext.define('OMV.module.admin.diagnostic.log.plugin.PhpFpm', {
    extend: 'OMV.module.admin.diagnostic.log.plugin.Plugin',
    alias: 'omv.plugin.diagnostic.log.phpfpm',

    id: 'php-fpm',
    text: 'PHP-FPM',
    columns: [{
        text: _('Date & Time'),
        sortable: true,
        dataIndex: 'rownum',
        stateId: 'date',
        renderer: function(value, metaData, record) {
            return record.get('date');
        }
    }, {
        text: _('Event'),
        sortable: true,
        dataIndex: 'event',
        stateId: 'event',
        flex: 1
    }],
    rpcParams: {
        id: 'php-fpm'
    },
    rpcFields: [{
        name: 'rownum',
        type: 'int'
    }, {
        name: 'ts',
        type: 'int'
    }, {
        name: 'date',
        type: 'string'
    }, {
        name: 'event',
        type: 'string'
    }]
});


var/www/openmediavault/js/omv/module/admin/service/nginx/window/Server.js
/**
 * Copyright (C) 2014-2017 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/window/Form.js")
// require("js/omv/workspace/window/plugin/ConfigObject.js")
// require("js/omv/form/plugin/LinkedFields.js")
// require("js/omv/form/field/SharedFolderComboBox.js")
// require("js/omv/form/field/UserComboBox.js")

Ext.define('OMV.module.admin.service.nginx.window.Server', {
    extend: 'OMV.workspace.window.Form',
    requires: [
        'OMV.workspace.window.plugin.ConfigObject',
        'OMV.form.plugin.LinkedFields',
        'OMV.form.field.SharedFolderComboBox',
        'OMV.form.field.UserComboBox'
    ],

    plugins: [{
        ptype: 'configobject'
    }, {
        ptype: 'linkedfields',
        correlations: [{
            name: [
                'use_root',
            ],
            conditions: [{
                name: 'sharedfolderref',
                op: 'z'
            }],
            properties: [
                'readOnly'
            ]
        }, {
            name: [
                'use_public_directory'
            ],
            conditions: [{
                name: 'use_root',
                value: true
            }, {
                name: 'sharedfolderref',
                op: 'n'
            }],
            properties: [
                '!readOnly'
            ]
        }, {
            name: [
                'public_directory'
            ],
            conditions: [{
                name: 'use_public_directory',
                value: true
            }],
            properties: [
                'show',
                '!allowBlank',
                '!readOnly'
            ]
        }, {
            name: [
                'server_name'
            ],
            conditions: [{
                name: 'host_type',
                value: 'name'
            }],
            properties: [
                'show',
                '!allowBlank',
                '!readOnly'
            ]
        }, {
            name: [
                'port',
                'port_default_server',
            ],
            conditions: [{
                name: 'ssl_force',
                value: true
            }],
            properties: [
                'allowBlank',
                'readOnly',
            ]
        }, {
            name: [
                'sslcertificateref',
                'ssl_port',
                'ssl_port_default_server',
                'ssl_force'
            ],
            conditions: [{
                name: 'ssl_enable',
                value: true
            }],
            properties: [
                'show',
                '!allowBlank',
                '!readOnly'
            ]
        }, {
            name: [
                'php_pool_ref',
                'php_use_default_config',
            ],
            conditions: [{
                name: 'php_enable',
                value: true
            }],
            properties: [
                '!allowBlank',
                '!readOnly',
                'show'
            ]
        }, {
            name: [
                'use_index_html',
                'use_index_php',
            ],
            conditions: [{
                name: 'use_index',
                value: true
            }],
            properties: [
                '!readOnly'
            ]
        }]
    }],

    rpcService: 'Nginx',
    rpcSetMethod: 'set',

    height: 600,
    hideResetButton: true,

    getFormItems: function() {
        return [{
            xtype: 'fieldset',
            title: _('General'),
            items: [{
                xtype: 'checkbox',
                name: 'enable',
                fieldLabel: _('Enable'),
                checked: true
            }, {
                xtype: 'sharedfoldercombo',
                name: 'sharedfolderref',
                fieldLabel: _('Directory'),
                allowBlank: true,
                allowNone: true,
                value: '',
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('The location needs to have at least read permissions for the user/group www-data.') + ' ' +
                        _('This option will populate the $root_path variable for usage in extra options or as document root.')
                }]
            }, {
                xtype: 'checkbox',
                name: 'use_root',
                fieldLabel: _('Use root'),
                checked: true,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('Use "Directory" as document root.')
                }]
            }, {
                xtype: 'checkbox',
                name: 'use_public_directory',
                fieldLabel: _('Use public directory'),
                checked: false,
                readOnly: true,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('Use if you serve your public files in a subfolder like public_html or public. This folder will be the document root. It also avoids the need to create an additional shared folder in such cases.')
                }]
            }, {
                xtype: 'combo',
                name: 'public_directory',
                fieldLabel: _('Public directory'),
                queryMode: 'local',
                store: Ext.create('Ext.data.ArrayStore', {
                    fields: [
                        'value'
                    ],
                    data: [
                        ['public_html'],
                        ['public'],
                    ]
                }),
                displayField: 'value',
                valueField: 'value',
                allowBlank: false,
                editable: true,
                triggerAction: 'all',
                value: 'public_html'
            }, {
                xtype: 'combo',
                name: 'host_type',
                fieldLabel: _('Host type'),
                queryMode: 'local',
                store: Ext.create('Ext.data.ArrayStore', {
                    fields: [
                        'value',
                        'text'
                    ],
                    data: [
                        ['port', _('Port')],
                        ['name', _('Name-based')]
                    ]
                }),
                displayField: 'text',
                valueField: 'value',
                allowBlank: false,
                editable: false,
                triggerAction: 'all',
                value: 'port'
            }, {
                xtype: 'textfield',
                name: 'server_name',
                fieldLabel: _('Server name'),
                allowBlank: true,
                readOnly: true,
                hidden: true
            }, {
                xtype: 'numberfield',
                name: 'port',
                fieldLabel: _('Port'),
                vtype: 'port',
                minValue: 0,
                maxValue: 65535,
                allowDecimals: false,
                allowNegative: false,
                value: 80
            }, {
                xtype: 'checkbox',
                name: 'port_default_server',
                fieldLabel: _('Default server'),
                checked: false,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('Set the server as the default one to be served if no server name matches on the selected port.')
                }]
            }]
        }, {
            xtype: 'fieldset',
            title: 'SSL',
            items: [{
                xtype: 'checkbox',
                name: 'ssl_enable',
                fieldLabel: _('Enable SSL'),
                checked: false,
                listeners: {
                    change: function(field, newValue) {
                        var sslForceField = this.findField('ssl_force');
                        sslForceField.setValue(false);
                    },
                    scope: this
                }
            }, {
                xtype: 'numberfield',
                name: 'ssl_port',
                fieldLabel: _('Port'),
                vtype: 'port',
                minValue: 0,
                maxValue: 65535,
                allowDecimals: false,
                allowNegative: false,
                value: 443
            }, {
                xtype: 'checkbox',
                name: 'ssl_port_default_server',
                fieldLabel: _('Default server'),
                checked: false,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('Set the server as the default one to be served if no server name matches on the selected port.')
                }]
            }, {
                xtype: 'sslcertificatecombo',
                name: 'sslcertificateref',
                fieldLabel: _('Certificate'),
                allowNone: true,
                allowBlank: true,
                readOnly: true,
                hidden: true,
                value: '',
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('The SSL certificate.')
                }]
            }, {
                xtype: 'checkbox',
                name: 'ssl_force',
                fieldLabel: _('Only use SSL'),
                checked: false
            }]
        }, {
            xtype: 'fieldset',
            title: 'PHP',
            items: [{
                xtype: 'checkbox',
                name: 'php_enable',
                fieldLabel: _('Enable PHP'),
                checked: false,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('This enables PHP and makes the $socket variable (for the chosen PHP-FPM pool) available in extra options.')
                }]
            }, {
                xtype: 'combo',
                name: 'php_pool_ref',
                fieldLabel: _('PHP-FPM pool'),
                emptyText: _('Select a PHP-FPM pool ...'),
                allowBlank: true,
                editable: false,
                displayField: 'name',
                valueField: 'uuid',
                store: Ext.create('OMV.data.Store', {
                    autoLoad: true,
                    model: OMV.data.Model.createImplicit({
                        identifier: 'empty',
                        idProperty: 'uuid',
                        fields: [{
                            name: 'uuid',
                            type: 'string'
                        }, {
                            name: 'name',
                            type: 'string'
                        }]
                    }),
                    proxy: {
                        type: 'rpc',
                        rpcData: {
                            service: 'PhpFpm',
                            method: 'getList'
                        },
                    },
                    sorters: [{
                        direction: 'ASC',
                        property: 'name'
                    }],
                    listeners: {
                        scope: this,
                        load: function(store) {
                            store.insert(0, {
                                uuid: '',
                                name: _('None')
                            });
                        }
                    }
                }),
                value: ''
            }, {
                xtype: 'checkbox',
                name: 'php_use_default_config',
                fieldLabel: _('Default config'),
                checked: true,
                plugins: [{
                    ptype: 'fieldinfo',
                    text: _('Use the default Nginx config for PHP.') + ' ' +
                        _('Uncheck this if you want to use your own configuration to connect to the PHP socket.')
                }]
            }]
        }, {
            xtype: 'fieldset',
            title: _('Options'),
            items: [{
                xtype: 'fieldset',
                title: _('Index'),
                items: [{
                    xtype: 'checkbox',
                    name: 'use_index',
                    boxLabel: _('Use index'),
                    checked: true,
                }, {
                    border: false,
                    layout: 'column',
                    defaults: {
                        border: false,
                        columnWidth: 0.5,
                        hideLabel: true,
                        labelSeparator: '',
                        layout: 'form'
                    },
                    items: [{
                        items: [{
                            xtype: 'checkbox',
                            name: 'use_index_html',
                            boxLabel: 'index.html',
                            checked: true
                        }]
                    }, {
                        items: [{
                            xtype: 'checkbox',
                            name: 'use_index_php',
                            boxLabel: 'index.php',
                            checked: false
                        }]
                    }]
                }]
            }, {
                border: false,
                layout: 'column',
                defaults: {
                    border: false,
                    columnWidth: 0.5,
                    hideLabel: true,
                    labelSeparator: '',
                    layout: 'form'
                },
                items: [{
                    items: [{
                        xtype: 'checkbox',
                        name: 'autoindex',
                        boxLabel: 'Autoindex',
                        checked: false
                    }, {
                        xtype: 'checkbox',
                        name: 'log_enable',
                        fieldLabel: _('Enable log'),
                        checked: true
                    }]
                }, {
                    items: [{
                        xtype: 'checkbox',
                        name: 'deny_htaccess',
                        boxLabel: _('Don\'t serve .htaccess'),
                        checked: false
                    }]
                }]
            }, {
                xtype: 'combo',
                name: 'large_client_header_buffers',
                fieldLabel: _('Header buffer size'),
                queryMode: 'local',
                store: Ext.create('Ext.data.ArrayStore', {
                    fields: [
                        'value',
                        'text'
                    ],
                    data: [
                        [8, '8k'],
                        [16, '16k'],
                        [32, '32k'],
                    ]
                }),
                displayField: 'text',
                valueField: 'value',
                allowBlank: false,
                editable: false,
                triggerAction: 'all',
                value: 8
            }]
        }, {
            xtype: 'fieldset',
            title: _('Extra options'),
            items: [{
                xtype: 'textarea',
                name: 'extra_options',
                grow: true,
                minHeight: 150,
                allowBlank: true
            }]
        }];
    }
});
